corehq/ex-submodules/soil/__init__.py

```diff
@@ -38,16 +38,30 @@ class DownloadBase(object):
 
     has_file = False
 
+    # `new_id_prefix` adopted when implementing new blob metadata
+    # format. This needs to round-trip to the client and back so we can
+    # detect legacy download ids, which should never start with
+    # `new_id_prefix`. Previously blobs were saved in the blob db using
+    # `self.identifier` as the key, which could result in collisions.
+    # Can be removed after all legacy blob downloads have expired.
+    new_id_prefix = "dl-"  # not a valid uuid4 prefix
+
     def __init__(self, mimetype="text/plain",
                  content_disposition='attachment; filename="download.txt"',
                  transfer_encoding=None, extras=None, download_id=None,
                  cache_backend=SOIL_DEFAULT_CACHE, content_type=None,
                  suffix=None, message=None):
+
+        if download_id is None:
+            download_id = self.new_id_prefix + uuid.uuid4().hex
+        else:
+            assert download_id.startswith(self.new_id_prefix), download_id
```

Comment:
is the class never created from an existing ID? What if a user accesses a download URL with an old ID - like from an 'export ready' email?

Reply:
> is the class never created from an existing ID?

If it is, the ID must start with `new_id_prefix`. This is only done in rare cases where `download_id` is needed before this object can be created. For example, when passing a download ID into a celery task that will use the key to save some content to be downloaded after it is finished.

> What if a user accesses a download URL with an old ID - like from an 'export ready' email?

The purpose of `new_id_prefix` is so we can tell the difference between old and new keys. Old downloads will not have this prefix and we use that fact to detect them and construct their blob key appropriately:

https://github.com/dimagi/commcare-hq/blob/64e62069396b00d4a13d4fdc8b79c3a358d2f7cd/corehq/ex-submodules/soil/__init__.py#L304-L309